### PR TITLE
Add support read hook and implement webAPI based read hook

### PR DESCRIPTION
**Pull request by hellais**
Based on discussion and feedback from @aidask in https://github.com/flowjs/flow.js/issues/42

### CODE REVIEW

**Comment by hellais** [comment]
I can't run the unittests locally so I am hoping this will kick off the travis unittest.

**Comment by AidasK** [comment]
test `FlowFile functions should get extension FAILED` - fails, because you have deleted getExtension method. It should not be hard to launch tests locally, you only need to install npm. Please ask if you have any serious questions.

**Comment by hellais** [comment]
I believe the above error is now fixed. What is the status on getting this merged into upstream?

**Comment by evilaliv3** [comment]
@hellais this patch is the only one needed by us right?

@AidasK can you provide us a feedback for the possibility to merge it and for a tentative release schedule?

**Comment by evilaliv3** [comment]
@AidasK can you provide us a feedback for the possibility to merge it and for a tentative release schedule?

File: src/flow.js [hunk 1]
```diff
@@ -1254,16 +1268,37 @@
...
+
+
+    /**
      * Uploads the actual data in a POST call
      * @function
      */
     send: function () {
-      var preprocess = this.flowObj.opts.preprocess;
+      var preprocess = this.flowObj.opts.preprocess,
+        read = this.flowObj.opts.read;
```
Comment (AidasK) [inline src/flow.js]: Is read defined?

File: src/flow.js [hunk 2]
```diff
@@ -1274,6 +1309,16 @@
             return;
         }
       }
+      if (typeof read === 'function') {
```
Comment (AidasK) [inline src/flow.js]: Can it be set to anything else? Maybe we do not need a `if` clause here.

File: src/flow.js [hunk 3]
```diff
@@ -1102,19 +1120,21 @@
...
+    this.startByte = this.offset * this.chunkSize;
 
     /**
      * Chunk end byte in a file
      * @type {number}
      */
-    this.endByte = Math.min(this.fileObjSize, (this.offset + 1) * chunkSize);
+    this.endByte = Math.min(this.fileObj.size, (this.offset + 1) * this.chunkSize);
+
+    this.data = null;
```
Comment (AidasK) [inline src/flow.js]: Is it used anywhere?

File: dist/flow.js [hunk 4]
```diff
@@ -28,6 +28,7 @@
    * @param {number} [opts.chunkRetryInterval]
```
Comment (AidasK) [inline dist/flow.js]: Could you delete `dist` files from a pull request?
dist/flow.min.js
dist/flow.js

**Comment by AidasK** [comment]
Could you also update documentation? It would be great to have an example, how you are using this for compression and encryption. Thanks!

File: src/flow.js (hunk 1 above)
Comment (evilaliv3) [inline src/flow.js]: yes it has the default you suggested:
```
      read: webAPIFileRead,
```

that is:
```
 /**
  * Default read function using the webAPI
  *
  * @function webAPIFileRead(chunk, startByte, endByte, fileType)
  *
  */
 function webAPIFileRead(chunk, startByte, endByte, fileType) {
   var function_name = (chunk.fileObj.file.slice ? 'slice' :
     (chunk.fileObj.file.mozSlice ? 'mozSlice' :
       (chunk.fileObj.file.webkitSlice ? 'webkitSlice' :
         'slice')));
   chunk.readFinished(chunk.fileObj.file[function_name](startByte, endByte, fileType));
 }
```

File: src/flow.js (hunk 2 above)
Comment (evilaliv3) [inline src/flow.js]: sure, i will modify this using the same code patterns you have already used.

File: src/flow.js (hunk 3 above)
Comment (evilaliv3) [inline src/flow.js]: let me recheck and if not i will remove it.

File: dist/flow.js (hunk 4 above)
Comment (evilaliv3) [inline dist/flow.js]: done.

File: src/flow.js (hunk 1 above)
Comment (AidasK) [inline src/flow.js]: read -> var read ?

File: src/flow.js [hunk 5]
```diff
@@ -89,9 +90,10 @@
       chunkRetryInterval: null,
       permanentErrors: [404, 415, 500, 501],
       successStatuses: [200, 201, 202],
-      onDropStopPropagation: false
+      onDropStopPropagation: false,
+      read: webAPIFileRead,
```
Comment (AidasK) [inline src/flow.js]: comma might cause a fata error for IE browsers

File: src/flow.js (hunk 2 above)
Comment (evilaliv3) [inline src/flow.js]: right. given that the read is always necessary (not like the process one that cna be set as null) we can simply remove the if. this way users that will set it wrong will notice an error that instead would be silently masked. given that currently flow.js does not validayte any opt i';m ok for removing the if.

File: src/flow.js (hunk 5 above)
Comment (evilaliv3) [inline src/flow.js]: k, i'm removing it.

File: src/flow.js (hunk 1 above)
Comment (evilaliv3) [inline src/flow.js]: an other time this is a code pattern (https://github.com/shichuan/javascript-patterns/blob/master/general-patterns/single-var-pattern.html) but given that you do not use it in other parts of the code i will change it.

Comment (AidasK) [inline src/flow.js]: Sorry, haven't noticed comma in line above. I just try to follow google style guide and to be consistent.

**Comment by evilaliv3** [comment]
this has been integrated a lot of time ago by means of a different pull request.

**Pull request closed without merging**